### PR TITLE
[WIP] Add apache license header to each file

### PR DESCRIPTION
**Pull request by mli**
TODO lists

- [ ] how to deal with the mkl license, see https://github.com/dmlc/mxnet/blob/master/src/operator/mkl/mkl_batch_norm-inl.h @glingyan 
- [ ] R-package license @thirdwing 
- [ ] Pass CI

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,3 +1,19 @@
+# or more contributor license agreements.  See the NOTICE file
```
Comment (bhavinthaker) [inline Makefile]: This line is missing:  # Licensed to the Apache Software Foundation (ASF) under one

Comment (mli) [inline Makefile]: good catch, i manually copied for this file.

**Comment by thirdwing** [comment]
@mli I think we don't need any special treatment for the R package. We are in the same situation with Spark.

In Spark R package (https://github.com/apache/spark/blob/master/R/pkg/src-native/string_hash_code.c), `R.h` and `Rinternals.h` have been included, which are both LGPL.

**Comment by felixcheung** [comment]
re R - we are talking on email. I don't think Spark would be a good reference - src-native is in .Rbuildignore and is not in any release.

**Comment by felixcheung** [comment]
for mkl, I think it should be kept with the original (still Apache, but not ASF) license header, and list in NOTICE file http://www.apache.org/legal/resolved.html#required-third-party-notices

**Comment by piiswrong** [comment]
@felixcheung What's the policy on copying code & doc snippets from other opensource projects?
Where should the license and attribution go?

**Comment by felixcheung** [comment]
@piiswrong this is described in [this link](http://www.apache.org/legal/resolved.html#required-third-party-notices)

first, it must not be in category x http://www.apache.org/legal/resolved.html#category-x

second, for any dependencies or copy of source not from an ASF project (note, this includes Apache 2.0 licenses)
- list it in LICENSE file in different license type groups eg. https://github.com/apache/spark/blob/master/LICENSE (notice how it referencing the licenses directory)
- for source included, reference the relative file path of each file and list them in the LICENSE file eg. https://github.com/apache/zeppelin/blob/master/LICENSE
- have a copy of its license under licenses directory eg. https://github.com/apache/spark/tree/master/licenses
- for any source include, its original header or licensing information is kept as-is (do not change to Apache/ASF header)

hope that helps

**Comment by piiswrong** [comment]
@felixcheung Thanks for the links.

Although this seems more like directions for submodules.
I'm referring to the cases where 10-20 lines of code is copied over and modified here and there.

**Comment by felixcheung** [comment]
@piiswrong hmm, that's usually very difficult to track down in case of any dispute. so generally should be discouraged, unless the contribution/commit/PR is made by the original owner of the code.

assuming we already code snippets taken from other projects in the source already, likely the best way is to attribute them as in:

> for source included, reference the relative file path of each file and list them in the LICENSE file eg. https://github.com/apache/zeppelin/blob/master/LICENSE

might be good to consult legal-discuss @ ASF on that too.

**Pull request closed without merging**